File: upstream/ray-gauges/packages/reactor-gen/src/accounts/ReactorConfig.ts
import { PublicKey, Connection } from "@solana/web3.js"
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface ReactorConfigFields {
  /** Token account that holds deposited RAY tokens */
  rayVault: PublicKey
  /**
   * Total RAY staked in all Reactors
   * Tracked separately here to avoid the vulnerability of tokens being sent directly to the vault
   */
  totalRayDeposited: BN
  /** Mint for the RAY token */
  rayMint: PublicKey
  /** Last time rewards were emitted */
  rewardsEmittedUntil: BN
  /** RAY per day */
  rayRewardDailyEmission: BN
  /** Token account that holds RAY tokens for emission */
  rayRewardHopper: PublicKey
  /** Increasing share index of RAY rewards */
  rayRewardIndex: types.NumberRawFields
  /** APR for isoRAY */
  isoRayAprBps: number
  /** isoRAY time-unit index */
  isoRayIndex: types.NumberRawFields
  /** Bump seed for the PDA */
  bump: Array<number>
}

export interface ReactorConfigJSON {
  /** Token account that holds deposited RAY tokens */
  rayVault: string
  /**
   * Total RAY staked in all Reactors
   * Tracked separately here to avoid the vulnerability of tokens being sent directly to the vault
   */
  totalRayDeposited: string
  /** Mint for the RAY token */
  rayMint: string
  /** Last time rewards were emitted */
  rewardsEmittedUntil: string
  /** RAY per day */
  rayRewardDailyEmission: string
  /** Token account that holds RAY tokens for emission */
  rayRewardHopper: string
  /** Increasing share index of RAY rewards */
  rayRewardIndex: types.NumberRawJSON
  /** APR for isoRAY */
  isoRayAprBps: number
  /** isoRAY time-unit index */
  isoRayIndex: types.NumberRawJSON
  /** Bump seed for the PDA */
  bump: Array<number>
}

export class ReactorConfig {
  /** Token account that holds deposited RAY tokens */
  readonly rayVault: PublicKey
  /**
   * Total RAY staked in all Reactors
   * Tracked separately here to avoid the vulnerability of tokens being sent directly to the vault
   */
  readonly totalRayDeposited: BN
  /** Mint for the RAY token */
  readonly rayMint: PublicKey
  /** Last time rewards were emitted */
  readonly rewardsEmittedUntil: BN
  /** RAY per day */
  readonly rayRewardDailyEmission: BN
  /** Token account that holds RAY tokens for emission */
  readonly rayRewardHopper: PublicKey
  /** Increasing share index of RAY rewards */
  readonly rayRewardIndex: types.NumberRaw
  /** APR for isoRAY */
  readonly isoRayAprBps: number
  /** isoRAY time-unit index */
  readonly isoRayIndex: types.NumberRaw
  /** Bump seed for the PDA */
  readonly bump: Array<number>

  static readonly discriminator = Buffer.from([
    45, 60, 88, 98, 145, 54, 88, 171,
  ])

  static readonly layout = borsh.struct([
    borsh.publicKey("rayVault"),
    borsh.u64("totalRayDeposited"),
    borsh.publicKey("rayMint"),
    borsh.u64("rewardsEmittedUntil"),
    borsh.u64("rayRewardDailyEmission"),
    borsh.publicKey("rayRewardHopper"),
    types.NumberRaw.layout("rayRewardIndex"),
    borsh.u16("isoRayAprBps"),
    types.NumberRaw.layout("isoRayIndex"),
    borsh.array(borsh.u8(), 1, "bump"),
  ])

  constructor(fields: ReactorConfigFields) {
    this.rayVault = fields.rayVault
    this.totalRayDeposited = fields.totalRayDeposited
    this.rayMint = fields.rayMint
    this.rewardsEmittedUntil = fields.rewardsEmittedUntil
    this.rayRewardDailyEmission = fields.rayRewardDailyEmission
    this.rayRewardHopper = fields.rayRewardHopper
    this.rayRewardIndex = new types.NumberRaw({ ...fields.rayRewardIndex })
    this.isoRayAprBps = fields.isoRayAprBps
    this.isoRayIndex = new types.NumberRaw({ ...fields.isoRayIndex })
    this.bump = fields.bump
  }

  static async fetch(
    c: Connection,
    address: PublicKey,
    programId: PublicKey = PROGRAM_ID
  ): Promise<ReactorConfig | null> {
    const info = await c.getAccountInfo(address)

    if (info === null) {
      return null
    }
    if (!info.owner.equals(programId)) {
      throw new Error("account doesn't belong to this program")
    }

    return this.decode(info.data)
  }

  static async fetchMultiple(
    c: Connection,
    addresses: PublicKey[],
    programId: PublicKey = PROGRAM_ID
  ): Promise<Array<ReactorConfig | null>> {
    const infos = await c.getMultipleAccountsInfo(addresses)

    return infos.map((info) => {
      if (info === null) {
        return null
      }
      if (!info.owner.equals(programId)) {
        throw new Error("account doesn't belong to this program")
      }

      return this.decode(info.data)
    })
  }

  static decode(data: Buffer): ReactorConfig {
    if (!data.slice(0, 8).equals(ReactorConfig.discriminator)) {
      throw new Error("invalid account discriminator")
    }

    const dec = ReactorConfig.layout.decode(data.slice(8))

    return new ReactorConfig({
      rayVault: dec.rayVault,
      totalRayDeposited: dec.totalRayDeposited,
      rayMint: dec.rayMint,
      rewardsEmittedUntil: dec.rewardsEmittedUntil,
      rayRewardDailyEmission: dec.rayRewardDailyEmission,
      rayRewardHopper: dec.rayRewardHopper,
      rayRewardIndex: types.NumberRaw.fromDecoded(dec.rayRewardIndex),
      isoRayAprBps: dec.isoRayAprBps,
      isoRayIndex: types.NumberRaw.fromDecoded(dec.isoRayIndex),
      bump: dec.bump,
    })
  }

  toJSON(): ReactorConfigJSON {
    return {
      rayVault: this.rayVault.toString(),
      totalRayDeposited: this.totalRayDeposited.toString(),
      rayMint: this.rayMint.toString(),
      rewardsEmittedUntil: this.rewardsEmittedUntil.toString(),
      rayRewardDailyEmission: this.rayRewardDailyEmission.toString(),
      rayRewardHopper: this.rayRewardHopper.toString(),
      rayRewardIndex: this.rayRewardIndex.toJSON(),
      isoRayAprBps: this.isoRayAprBps,
      isoRayIndex: this.isoRayIndex.toJSON(),
      bump: this.bump,
    }
  }

  static fromJSON(obj: ReactorConfigJSON): ReactorConfig {
    return new ReactorConfig({
      rayVault: new PublicKey(obj.rayVault),
      totalRayDeposited: new BN(obj.totalRayDeposited),
      rayMint: new PublicKey(obj.rayMint),
      rewardsEmittedUntil: new BN(obj.rewardsEmittedUntil),
      rayRewardDailyEmission: new BN(obj.rayRewardDailyEmission),
      rayRewardHopper: new PublicKey(obj.rayRewardHopper),
      rayRewardIndex: types.NumberRaw.fromJSON(obj.rayRewardIndex),
      isoRayAprBps: obj.isoRayAprBps,
      isoRayIndex: types.NumberRaw.fromJSON(obj.isoRayIndex),
      bump: obj.bump,
    })
  }
}

File: upstream/ray-gauges/sol/programs/reactor/src/state/reactor_config.rs
use crate::{state::NumberRaw, REACTOR_CONFIG_SEED};
use anchor_lang::prelude::*;
use precise_number::Number;

use super::Rewarder;

const SECONDS_IN_DAY: u64 = 86_400;
const SECONDS_IN_YEAR: u64 = SECONDS_IN_DAY * 365;

#[derive(Default)]
#[account]
pub struct ReactorConfig {
    /// Token account that holds deposited RAY tokens
    pub ray_vault: Pubkey,

    /// Total RAY staked in all Reactors
    /// Tracked separately here to avoid the vulnerability of tokens being sent directly to the vault
    pub total_ray_deposited: u64,

    /// Mint for the RAY token
    pub ray_mint: Pubkey,

    /// Last time rewards were emitted
    pub rewards_emitted_until: u64,

    /// RAY per day
    pub ray_reward_daily_emission: u64,

    /// Token account that holds RAY tokens for emission
    pub ray_reward_hopper: Pubkey,

    /// Increasing share index of RAY rewards
    pub ray_reward_index: NumberRaw,

    /// APR for isoRAY
    pub iso_ray_apr_bps: u16,

    /// isoRAY time-unit index
    pub iso_ray_index: NumberRaw,

    /// Bump seed for the PDA
    pub bump: [u8; 1],
}

impl ReactorConfig {
    pub const LEN: usize =
        // discriminator
        8 +

        // ray_vault
        32 +

        // total_ray_deposited
        8 +

        // ray_mint
        32 +

        // rewards_emitted_until
        8 +

        // ray_reward_daily_emission
        8 +

        // ray_reward_hopper
        32 +

        // ray_reward_index
        NumberRaw::LEN +

        // iso_ray_apr_bps
        2 +

        // iso_ray_index
        NumberRaw::LEN +

        // bump
        1;

    pub fn seeds(&self) -> [&[u8]; 2] {
        [REACTOR_CONFIG_SEED.as_bytes(), &self.bump]
    }

    pub fn deposit_ray(&mut self, amount: u64, current_ts: u64) {
        self.accrue_rewards(current_ts);

        self.total_ray_deposited = self
            .total_ray_deposited
            .checked_add(amount)
            .expect("total ray deposited overflow");
    }

    pub fn withdraw_ray(&mut self, amount: u64, current_ts: u64) {
        self.accrue_rewards(current_ts);

        self.total_ray_deposited = self
            .total_ray_deposited
            .checked_sub(amount)
            .expect("total ray deposited underflow");
    }

    fn accrue_rewards(&mut self, current_ts: u64) {
        if self.rewards_emitted_until >= current_ts {
            msg!("Time has not passed, skipping");
            return;
        }

        let time_elapsed = current_ts - self.rewards_emitted_until;

        self.emit_ray_reward(time_elapsed);
        self.accrue_iso_ray(time_elapsed);

        self.rewards_emitted_until = current_ts;
    }

    /// Emit RAY rewards by increasing the share index
    fn emit_ray_reward(&mut self, time_elapsed: u64) {
        // no need to emit rewards if no time has passed
        if time_elapsed == 0 {
            return;
        }

        // total shares
        let total_shares = self.total_ray_deposited;

        // if there are no shares, no need to increase the share index
        if total_shares == 0 {
            return;
        }

        // duration as a ratio of a day
        let duration_day = Number::from_ratio(time_elapsed.into(), SECONDS_IN_DAY.into());

        // RAY emission
        let ray_emission = Number::from_natural_u64(self.ray_reward_daily_emission) * duration_day;

        let total_shares = Number::from_natural_u64(total_shares);

        // calculate RAY per share
        let ray_per_share_increase = ray_emission / total_shares;

        // increase the share index
        let cur_ray_reward_index: Number = self.ray_reward_index.into();
        let new_ray_reward_index = cur_ray_reward_index + ray_per_share_increase;

        self.ray_reward_index = new_ray_reward_index.into();
    }

    /// Accrue isoRAY for the duration since the last accrual
    /// Increase both:
    /// - total isoRAY in circulation
    /// - isoRAY index (share value of RAY to isoRAY)
    fn accrue_iso_ray(&mut self, time_elapsed: u64) {
        if self.total_ray_deposited == 0 {
            msg!("no ray deposited, skipping accrue");
            return;
        }

        let duration_year = Number::from_ratio(time_elapsed.into(), SECONDS_IN_YEAR.into());
        let iso_ray_apr = Number::from_bps(self.iso_ray_apr_bps);

        let partial_rate = duration_year * iso_ray_apr;

        self.iso_ray_index += partial_rate;
    }
}

impl Rewarder for ReactorConfig {
    fn ray_reward_index(&self) -> Number {
        self.ray_reward_index.into()
    }

    fn iso_ray_index(&self) -> Number {
        self.iso_ray_index.into()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_emit_ray_reward() {
        let mut reactor_config = ReactorConfig::default();
        reactor_config.ray_reward_daily_emission = 100_000;

        // 0 RAY deposited, 0 time elapsed
        reactor_config.emit_ray_reward(0);
        assert_eq!(Number::ZERO, reactor_config.ray_reward_index.into());

        // 0 RAY deposited, 1 day elapsed
        reactor_config.emit_ray_reward(SECONDS_IN_DAY);
        assert_eq!(Number::ZERO, reactor_config.ray_reward_index.into());

        // 1 RAY deposited, 1 day elapsed
        // With 100_000 RAY daily emission, the share index should be 100_000
        reactor_config.total_ray_deposited = 1;
        reactor_config.emit_ray_reward(SECONDS_IN_DAY);
        assert_eq!(
            Number::from_natural_u64(100_000),
            reactor_config.ray_reward_index.into()
        );

        // now with 2 RAY deposited, the share index should increase by 50_000
        reactor_config.total_ray_deposited = 2;
        reactor_config.emit_ray_reward(SECONDS_IN_DAY);
        assert_eq!(
            Number::from_natural_u64(150_000),
            reactor_config.ray_reward_index.into()
        );
    }

    #[test]
    fn test_accrue_iso_ray() {
        let mut reactor_config = ReactorConfig::default();
        // 50% APR
        reactor_config.iso_ray_apr_bps = 50_00;

        // 0 RAY deposited, 0 time elapsed
        reactor_config.accrue_iso_ray(0);
        assert_eq!(Number::ZERO, reactor_config.iso_ray_index.into());

        // 1 RAY deposited, 0 time elapsed
        reactor_config.total_ray_deposited = 1;
        reactor_config.accrue_iso_ray(0);
        assert_eq!(Number::ZERO, reactor_config.iso_ray_index.into());

        // 1 RAY deposited, 1 year elapsed
        // With 50% APR, the isoRAY index should be 0.5 -- or 0.5 isoRAY per RAY
        reactor_config.accrue_iso_ray(SECONDS_IN_YEAR);
        assert_eq!(
            Number::from_ratio(1, 2),
            reactor_config.iso_ray_index.into()
        );

        // 2 RAY deposited, 1 year elapsed
        // After 1 year, 1 full isoRAY should be minted
        // This is 1 isoRAY divided by 2 RAY, or 0.5 isoRAY per RAY
        reactor_config.total_ray_deposited = 2;
        reactor_config.accrue_iso_ray(SECONDS_IN_YEAR);
        assert_eq!(Number::ONE, reactor_config.iso_ray_index.into());
    }
}
